instrumentation/github.com/go-gorm/gorm/otelgorm/callbacks.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelgorm
+
+import (
+	"strings"
+
+	"gorm.io/gorm"
+
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/semconv"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+const (
+	spanName = "gorm.query"
+
+	dbTableKey     = label.Key("db.table")
+	dbCountKey     = label.Key("db.count")
+	dbOperationKey = semconv.DBOperationKey
+	dbStatementKey = semconv.DBStatementKey
+)
+
+func dbTable(name string) label.KeyValue {
+	return dbTableKey.String(name)
+}
+
+func dbStatement(stmt string) label.KeyValue {
+	return dbStatementKey.String(stmt)
+}
+
+func dbCount(n int64) label.KeyValue {
+	return dbCountKey.Int64(n)
+}
+
+func dbOperation(op string) label.KeyValue {
+	return dbOperationKey.String(op)
+}
+
+func (op *OtelPlugin) before(tx *gorm.DB) {
+	tx.Statement.Context, _ = op.tracer.
+		Start(tx.Statement.Context, spanName, oteltrace.WithSpanKind(oteltrace.SpanKindClient))
```

Comment:
Should we align the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md) by changing the span name?

Reply:
not sure i understand the question

Reply:
> SQL spans SHOULD be named the following way, unless the statement is known to be of low cardinality: <db.operation> <db.name>.<db.sql.table>, provided that db.operation and db.sql.table are available.

For instance, a span name of a `SELECT` call might be `SELECT ShopDb.orders`; `ShopDb` is the database name, `orders` is the table.

> i saw that mongodb uses similar approach of mongodb.query

Good catch, I raised a [comment](https://github.com/open-telemetry/opentelemetry-go-contrib/issues/460#issuecomment-775812136) to address that.

Reply:
The table name is not obvious or easily accessible in many use cases.
What if the select is performed with a join over several tables? I'm not sure that convention makes much sense.
In order to achieve the that specification, one must parse the SQL query and understanding the SQL syntax which is beyond the scope of what this plugin is meant to solve.

Reply:
> The table name is not obvious or easily accessible in many use cases.

I found table can be accessed from `gorm.DB.Statement.Table`.

> What if the select is performed with a join over several tables? In order to achieve that specification, one must parse the SQL query and understanding the SQL syntax which is beyond the scope of what this plugin is meant to solve.

Then we should ignore the table, just `<db.operation> <db.name>`. Similarly, if the database name is not available, we should unset this field, just `<db.operation>`.

From the specification:

> If the operation is acting upon an anonymous table, or more than one table, this value MUST NOT be set.

> It is not recommended to attempt any client-side parsing of `db.statement`.

-----

I'd keep the span name in low cardinality but **distinct**, which can describe an "interesting class of spans".

Reply:
`gorm.DB.Statement.Table` is not always available. 
I will do best effort resolving for that.